File: test/allreduce_bottom.f90
program allreduce_bottom_example
    use mpi
    implicit none

    integer :: ierr, rank, size, i
    integer, dimension(:), allocatable :: my_values, result_values
    integer(kind=MPI_ADDRESS_KIND) :: disp_send, disp_recv
    integer(kind=MPI_ADDRESS_KIND), dimension(2) :: displacements
    integer, dimension(2) :: block_lengths
    integer :: datatype
    integer(kind=MPI_ADDRESS_KIND) :: base_address

    call MPI_Init(ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)
    call MPI_Comm_size(MPI_COMM_WORLD, size, ierr)

    ! Each process has a unique array of values
    allocate(my_values(size))
    allocate(result_values(size))

    ! Initialize the arrays
    my_values = rank + 1
    result_values = 0

    ! Determine the base address of memory
    call MPI_Get_address(my_values, base_address, ierr)
    call MPI_Get_address(my_values(1), disp_send, ierr)
    call MPI_Get_address(result_values, disp_recv, ierr)

    ! Calculate displacements relative to MPI_BOTTOM
    displacements(1) = disp_send - base_address
    displacements(2) = disp_recv - base_address

    block_lengths = [size, size]  ! The block sizes for both buffers

    ! Create a derived datatype for the operation
    call MPI_Type_create_hindexed(2, block_lengths, displacements, MPI_INTEGER, datatype, ierr)
    call MPI_Type_commit(datatype, ierr)

    ! Perform the reduction operation using MPI_BOTTOM
    call MPI_Allreduce(MPI_BOTTOM, MPI_BOTTOM, 1, datatype, MPI_SUM, MPI_COMM_WORLD, ierr)

    ! Check the result
    print *, 'Process', rank, 'my_values:', my_values
    print *, 'Process', rank, 'result_values after MPI_Allreduce with MPI_BOTTOM:', result_values

    ! Clean up
    call MPI_Type_free(datatype, ierr)
    deallocate(my_values)
    deallocate(result_values)
    call MPI_Finalize(ierr)
end program allreduce_bottom_example

File: test/allreduce_bottom.cpp
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>


#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

// Custom reduction function
void custom_sum_function(void *invec, void *inoutvec, int *len, MPI_Datatype *datatype) {
    int *in = (int *)invec;
    int *inout = (int *)inoutvec;
    for (int i = 0; i < *len; i++) {
        inout[i] += in[i];
    }
}

int main(int argc, char *argv[]) {
    int rank, size, ierr;
    int *send_buffer, *recv_buffer;
    MPI_Aint base_address, recv_disp;
    MPI_Aint displacements[1];
    int block_lengths[1];
    MPI_Datatype recv_type;
    MPI_Op custom_op;

    // Initialize MPI
    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    // Allocate and initialize buffers
    send_buffer = (int *)malloc(size * sizeof(int));
    recv_buffer = (int *)malloc(size * sizeof(int));

    for (int i = 0; i < size; i++) {
        send_buffer[i] = rank + 1; // Each process has unique values
        recv_buffer[i] = 0;        // Initialize receive buffer
    }

    // Get base address of recv_buffer
    MPI_Get_address(recv_buffer, &base_address);
    MPI_Get_address(&recv_buffer[0], &recv_disp);

    // Calculate displacement for recv_buffer relative to base address
    displacements[0] = recv_disp - base_address;

    // Define block length
    block_lengths[0] = size; // Number of integers in recv_buffer

    // Create a custom MPI datatype for the receive buffer
    MPI_Type_create_hindexed(1, block_lengths, displacements, MPI_INT, &recv_type);
    MPI_Type_commit(&recv_type);

    // Create a custom reduction operation
    MPI_Op_create(custom_sum_function, /*commute=*/1, &custom_op);

    // Perform an Allreduce operation using MPI_BOTTOM only for recv_buffer
    MPI_Allreduce(send_buffer, MPI_BOTTOM, 1, recv_type, custom_op, MPI_COMM_WORLD);

    // Print results
    printf("Process %d - send_buffer: ", rank);
    for (int i = 0; i < size; i++) {
        printf("%d ", send_buffer[i]);
    }
    printf("\n");

    printf("Process %d - recv_buffer after MPI_Allreduce: ", rank);
    for (int i = 0; i < size; i++) {
        printf("%d ", recv_buffer[i]);
    }
    printf("\n");

    // Clean up
    MPI_Type_free(&recv_type);
    MPI_Op_free(&custom_op);
    free(send_buffer);
    free(recv_buffer);
    MPI_Finalize();

    return 0;
}

